src/memory-analyzer/analyze_symbol.cpp

```diff
@@ -0,0 +1,476 @@
+    }
+  }
+  return array;
+}
+
+exprt symbol_analyzert::fill_expr_with_values(
+  const std::string &symbol_id,
+  exprt expr,
+  const typet &type,
+  const source_locationt &location)
+{
+  if(expr.is_constant() && (is_c_int_derivate(type) || is_c_char(type)))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    const mp_integer int_rep = string2integer(value, DECIMAL_SYSTEM);
+    return from_integer(int_rep, type);
+  }
+  else if(expr.is_constant() && is_c_bool(type))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    return from_c_boolean_value(value, type);
+  }
+  else if(expr.is_constant() && is_c_enum(expr.type()))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    return convert_memeber_name_to_enum_value(
+      value, to_c_enum_type(expr.type()));
+  }
+  else if(is_struct(type))
+  {
+    return fill_struct_with_values(symbol_id, expr, type, location);
+  }
+  else if(is_array(type))
+  {
+    expr = fill_array_with_values(symbol_id, expr, type, location);
+  }
+  else
+  {
+    throw symbol_analysis_exceptiont(
+      "unexpected thing:  " + expr.pretty() + "\nexception end\n");
+  }
+  return expr;
+}
+
+exprt symbol_analyzert::fill_char_struct_member_with_values(
+  const symbol_exprt &field_access,
+  const exprt &default_expr,
+  const source_locationt &location)
+{
+  const std::string memory_location =
+    gdb_api.get_memory(id2string(field_access.get_identifier()));
+  if(memory_location != "0x0")
+  {
+    return declare_and_initalize_char_ptr(
+      field_access, memory_location, location);
+  }
+  return default_expr;
+}
+
+exprt symbol_analyzert::fill_struct_with_values(
+  const std::string &symbol_id,
+  exprt &expr,
+  const typet &type,
+  const source_locationt &location)
+{
+  const struct_typet structt = to_struct_type(type);
+  for(size_t i = 0; i < expr.operands().size(); ++i)
+  {
+    exprt &operand = expr.operands()[i];
+    const typet &resolved_type = ns.follow(operand.type());
+    if(!structt.components()[i].get_is_padding())
+    {
+      std::string field_access =
+        symbol_id + "." + id2string(structt.components()[i].get_name());
+      symbol_exprt field_symbol(field_access, resolved_type);
+
+      if(is_c_char_pointer(resolved_type))
+      {
+        operand =
+          fill_char_struct_member_with_values(field_symbol, operand, location);
+      }
+      else if(is_void_pointer(resolved_type))
+      {
+        operand = fill_void_pointer_with_values(field_symbol, location);
+      }
+      else if(is_struct(resolved_type))
+      {
+        code_declt declaration = declare_instance(resolved_type);
+
+        const exprt target_expr = fill_expr_with_values(
+          field_access,
+          zero_initializer(resolved_type, location, ns),
+          resolved_type,
+          location);
+
+        declaration.operands().resize(2);
+        declaration.op1() = target_expr;
```

Comment:
This is a recurring pattern - maybe `declare_instance` should just take a second parameter?